problem-4/problem-4.test.js

```diff
@@ -1,21 +1,70 @@
-const solution = () => {
+//1. 가장 익숙한 방법으로 문제를 해결해 주세요.
+{
+  /*
+const solution = (n) => {
+  let decimal = 0;
+  let exponent = 0;
+  if (n === "0" || n === "1") {
+    return parseInt(n, 10);
+  }
+  const length = n.length;
+  while (exponent <= length - 1) {
+    decimal += parseInt(n.charAt(n.length - 1), 10) * 2 ** exponent;
+
+    exponent++;
+    n = n.slice(0, -1);
+  }
+  return decimal;
+};*/
+}
+//2. 이번에는 재귀 함수로 문제를 해결해 주세요.
+{
+  /*
+const solution = (n, acc = 0, exponent = 0) => {
+  if (n.length === 1) {
+    return acc + parseInt(n, 10) * 2 ** exponent;
+  }
+  acc += parseInt(n.charAt(n.length - 1), 10) * 2 ** exponent;
+  return solution(n.slice(0, -1), acc, exponent + 1);
+};*/
+}
+
+//3. 꼬리 재귀 함수로 바꿔보세요.
+//4. 꼬리 재귀 최적화를 통해서 최적화해 보세요.
+const solution = (n) => {
+  let acc = 0;
+  let exponent = 0;
+
+  const length = n.length;
+  if (length === 1) {
+    return acc + parseInt(n, 10) * 2 ** exponent;
```

Comment:
연산자의 우선순위는 프로그래밍 언어마다 다를수도 있어요. 그래서 a + b + c같이 아주 자명한 경우가 아니라면 적절히 괄호로 묶어주는 것이 좋습니다.

```js
    return acc + parseInt(n, 10) * (2 ** exponent);
```
```suggestion
    return acc + parseInt(n, 10) * 2 ** exponent;
```